File: Medium/341. Flatten Nested List Iterator/soln.java
import java.util.*;

public class NestedIterator {
    private List<Integer> flatList;
    private int current;

    public NestedIterator(List<NestedInteger> nestedList) {
        flatList = new ArrayList<>();
        current = 0;
        flatten(nestedList);
    }

    public int next() {
        int result = flatList.get(current);
        current++;
        return result;
    }

    public boolean hasNext() {
        return current < flatList.size();
    }

    private void flatten(List<NestedInteger> nestedList) {
        for (NestedInteger item : nestedList) {
            if (item.isInteger()) {
                flatList.add(item.getInteger());
            } else {
                flatten(item.getList());
            }
        }
    }
}

File: Medium/341. Flatten Nested List Iterator/soln.cpp
class NestedIterator {
public:
    NestedIterator(vector<NestedInteger> &nestedList) {
        flatten(nestedList);
        current = 0;
    }

    int next() {
        return flatList[current++];
    }

    bool hasNext() {
        return current < flatList.size();
    }

private:
    vector<int> flatList;
    int current;

    void flatten(const vector<NestedInteger>& nestedList) {
        for (const auto &item : nestedList) {
            if (item.isInteger()) {
                flatList.push_back(item.getInteger());
            } else {
                flatten(item.getList());
            }
        }
    }
};
